From f30de61fc392376c5535b03ee9b0bca6c508fc8a Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Tue, 16 Sep 2014 17:01:19 -0400
Subject: [usb] ehci: support running URB giveback in tasklet context""

Message-id: <1410887463-20674-8-git-send-email-dzickus@redhat.com>
Patchwork-id: 95100
O-Subject: [RHEL7 PATCH 007/591] Revert "Revert "USB: EHCI: support running URB giveback in tasklet context""
Bugzilla: 1110939
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1110939

commit c04ee4b1136e462722567cf6e76bb35a181574a7
Author: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date:   Mon Sep 23 13:32:51 2013 -0700

    Revert "Revert "USB: EHCI: support running URB giveback in tasklet context""

    This reverts commit 3b8d7321ed4b8511e17048303b806ffcc2806077, which
    brings back commit 428aac8a81058e2303677a8fbf26670229e51d3a as it should
    be working for the 3.13-rc1 merge window now that Alan's other fixes are
    here in the tree already.

    Cc: Alan Stern <stern@rowland.harvard.edu>
    Cc: Ming Lei <ming.lei@canonical.com>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/usb/host/ehci-fsl.c b/drivers/usb/host/ehci-fsl.c
index 62f81f1..38852d0 100644
--- a/drivers/usb/host/ehci-fsl.c
+++ b/drivers/usb/host/ehci-fsl.c
@@ -663,7 +663,7 @@ static const struct hc_driver ehci_fsl_hc_driver = {
   * generic hardware linkage
   */
  .irq = ehci_irq,
- .flags = HCD_USB2 | HCD_MEMORY,
+ .flags = HCD_USB2 | HCD_MEMORY | HCD_BH,
 
  /*
   * basic lifecycle operations
diff --git a/drivers/usb/host/ehci-grlib.c b/drivers/usb/host/ehci-grlib.c
index 5d75de9..c6048ee 100644
--- a/drivers/usb/host/ehci-grlib.c
+++ b/drivers/usb/host/ehci-grlib.c
@@ -43,7 +43,7 @@ static const struct hc_driver ehci_grlib_hc_driver = {
   * generic hardware linkage
   */
  .irq   = ehci_irq,
- .flags   = HCD_MEMORY | HCD_USB2,
+ .flags   = HCD_MEMORY | HCD_USB2 | HCD_BH,
 
  /*
   * basic lifecycle operations
diff --git a/drivers/usb/host/ehci-hcd.c b/drivers/usb/host/ehci-hcd.c
index b0b26b9..069f63f 100644
--- a/drivers/usb/host/ehci-hcd.c
+++ b/drivers/usb/host/ehci-hcd.c
@@ -1158,7 +1158,7 @@ static const struct hc_driver ehci_hc_driver = {
   * generic hardware linkage
   */
  .irq =   ehci_irq,
- .flags =  HCD_MEMORY | HCD_USB2,
+ .flags =  HCD_MEMORY | HCD_USB2 | HCD_BH,
 
  /*
   * basic lifecycle operations
diff --git a/drivers/usb/host/ehci-mv.c b/drivers/usb/host/ehci-mv.c
index 4020629..6516d24 100644
--- a/drivers/usb/host/ehci-mv.c
+++ b/drivers/usb/host/ehci-mv.c
@@ -96,7 +96,7 @@ static const struct hc_driver mv_ehci_hc_driver = {
   * generic hardware linkage
   */
  .irq = ehci_irq,
- .flags = HCD_MEMORY | HCD_USB2,
+ .flags = HCD_MEMORY | HCD_USB2 | HCD_BH,
 
  /*
   * basic lifecycle operations
diff --git a/drivers/usb/host/ehci-octeon.c b/drivers/usb/host/ehci-octeon.c
index a89750f..4b76eb0 100644
--- a/drivers/usb/host/ehci-octeon.c
+++ b/drivers/usb/host/ehci-octeon.c
@@ -51,7 +51,7 @@ static const struct hc_driver ehci_octeon_hc_driver = {
   * generic hardware linkage
   */
  .irq   = ehci_irq,
- .flags   = HCD_MEMORY | HCD_USB2,
+ .flags   = HCD_MEMORY | HCD_USB2 | HCD_BH,
 
  /*
   * basic lifecycle operations
diff --git a/drivers/usb/host/ehci-pmcmsp.c b/drivers/usb/host/ehci-pmcmsp.c
index 601e208..893b707 100644
--- a/drivers/usb/host/ehci-pmcmsp.c
+++ b/drivers/usb/host/ehci-pmcmsp.c
@@ -286,7 +286,7 @@ static const struct hc_driver ehci_msp_hc_driver = {
 #else
  .irq =   ehci_irq,
 #endif
- .flags =  HCD_MEMORY | HCD_USB2,
+ .flags =  HCD_MEMORY | HCD_USB2 | HCD_BH,
 
  /*
   * basic lifecycle operations
diff --git a/drivers/usb/host/ehci-ppc-of.c b/drivers/usb/host/ehci-ppc-of.c
index 56dc732..da95a31 100644
--- a/drivers/usb/host/ehci-ppc-of.c
+++ b/drivers/usb/host/ehci-ppc-of.c
@@ -28,7 +28,7 @@ static const struct hc_driver ehci_ppc_of_hc_driver = {
   * generic hardware linkage
   */
  .irq   = ehci_irq,
- .flags   = HCD_MEMORY | HCD_USB2,
+ .flags   = HCD_MEMORY | HCD_USB2 | HCD_BH,
 
  /*
   * basic lifecycle operations
diff --git a/drivers/usb/host/ehci-ps3.c b/drivers/usb/host/ehci-ps3.c
index fd98377..8188542 100644
--- a/drivers/usb/host/ehci-ps3.c
+++ b/drivers/usb/host/ehci-ps3.c
@@ -71,7 +71,7 @@ static const struct hc_driver ps3_ehci_hc_driver = {
  .product_desc  = "PS3 EHCI Host Controller",
  .hcd_priv_size  = sizeof(struct ehci_hcd),
  .irq   = ehci_irq,
- .flags   = HCD_MEMORY | HCD_USB2,
+ .flags   = HCD_MEMORY | HCD_USB2 | HCD_BH,
  .reset   = ps3_ehci_hc_reset,
  .start   = ehci_run,
  .stop   = ehci_stop,
diff --git a/drivers/usb/host/ehci-q.c b/drivers/usb/host/ehci-q.c
index a7f776a..e321804 100644
--- a/drivers/usb/host/ehci-q.c
+++ b/drivers/usb/host/ehci-q.c
@@ -247,8 +247,6 @@ static int qtd_copy_status (
 
 static void
 ehci_urb_done(struct ehci_hcd *ehci, struct urb *urb, int status)
-__releases(ehci->lock)
-__acquires(ehci->lock)
 {
  if (usb_pipetype(urb->pipe) == PIPE_INTERRUPT) {
   /* ... update hc-wide periodic stats */
@@ -274,11 +272,8 @@ __acquires(ehci->lock)
   urb->actual_length, urb->transfer_buffer_length);
 #endif
 
- /* complete() can reenter this HCD */
  usb_hcd_unlink_urb_from_ep(ehci_to_hcd(ehci), urb);
- spin_unlock (&ehci->lock);
  usb_hcd_giveback_urb(ehci_to_hcd(ehci), urb, status);
- spin_lock (&ehci->lock);
 }
 
 static int qh_schedule (struct ehci_hcd *ehci, struct ehci_qh *qh);
diff --git a/drivers/usb/host/ehci-sead3.c b/drivers/usb/host/ehci-sead3.c
index f55477c..9877870 100644
--- a/drivers/usb/host/ehci-sead3.c
+++ b/drivers/usb/host/ehci-sead3.c
@@ -55,7 +55,7 @@ const struct hc_driver ehci_sead3_hc_driver = {
   * generic hardware linkage
   */
  .irq   = ehci_irq,
- .flags   = HCD_MEMORY | HCD_USB2,
+ .flags   = HCD_MEMORY | HCD_USB2 | HCD_BH,
 
  /*
   * basic lifecycle operations
diff --git a/drivers/usb/host/ehci-sh.c b/drivers/usb/host/ehci-sh.c
index b44d716..50e665d 100644
--- a/drivers/usb/host/ehci-sh.c
+++ b/drivers/usb/host/ehci-sh.c
@@ -36,7 +36,7 @@ static const struct hc_driver ehci_sh_hc_driver = {
   * generic hardware linkage
   */
  .irq    = ehci_irq,
- .flags    = HCD_USB2 | HCD_MEMORY,
+ .flags    = HCD_USB2 | HCD_MEMORY | HCD_BH,
 
  /*
   * basic lifecycle operations
diff --git a/drivers/usb/host/ehci-tilegx.c b/drivers/usb/host/ehci-tilegx.c
index b083a35..2c364aa 100644
--- a/drivers/usb/host/ehci-tilegx.c
+++ b/drivers/usb/host/ehci-tilegx.c
@@ -61,7 +61,7 @@ static const struct hc_driver ehci_tilegx_hc_driver = {
   * Generic hardware linkage.
   */
  .irq   = ehci_irq,
- .flags   = HCD_MEMORY | HCD_USB2,
+ .flags   = HCD_MEMORY | HCD_USB2 | HCD_BH,
 
  /*
   * Basic lifecycle operations.
diff --git a/drivers/usb/host/ehci-w90x900.c b/drivers/usb/host/ehci-w90x900.c
index 59e0e24..1c370df 100644
--- a/drivers/usb/host/ehci-w90x900.c
+++ b/drivers/usb/host/ehci-w90x900.c
@@ -108,7 +108,7 @@ static const struct hc_driver ehci_w90x900_hc_driver = {
   * generic hardware linkage
   */
  .irq = ehci_irq,
- .flags = HCD_USB2|HCD_MEMORY,
+ .flags = HCD_USB2|HCD_MEMORY|HCD_BH,
 
  /*
   * basic lifecycle operations
diff --git a/drivers/usb/host/ehci-xilinx-of.c b/drivers/usb/host/ehci-xilinx-of.c
index d845e3b..bb96a68 100644
--- a/drivers/usb/host/ehci-xilinx-of.c
+++ b/drivers/usb/host/ehci-xilinx-of.c
@@ -79,7 +79,7 @@ static const struct hc_driver ehci_xilinx_of_hc_driver = {
   * generic hardware linkage
   */
  .irq   = ehci_irq,
- .flags   = HCD_MEMORY | HCD_USB2,
+ .flags   = HCD_MEMORY | HCD_USB2 | HCD_BH,
 
  /*
   * basic lifecycle operations
-- 
1.7.1